### ISSUE TITLE
Streamlined bioconda recipe development

### ISSUE THREAD

**Issue by daler** [bioconda/bioconda-recipes#5374]
@bioconda/all we have some new changes that should hopefully make it easier to build and test recipes locally.

## Summary

`simulate-travis.py` will now install a completely isolated conda environment exclusively for its own use. 

To start using it, first perform the setup (putting in whatever path you'd like to install the isolated version in):

```bash
./simulate-travis.py --bootstrap /tmp/miniconda 
```

That's it. You can then use `simulate-travis.py` as before. You can even remove `bioconda-utils` and `conda-build` from your root environment. Some folks use `bioconda-utils` directly, in which case you would specify the full path `/tmp/miniconda/bin/bioconda-utils`.

There are some additional features as well, see below.

## Rationale

Previously, contributors were forced to change their root conda environment to run tests locally. This is because `bioconda-utils` requires `conda-build`, and `conda-build` must be installed in the root environment. This was frustrating if the root environment was used for something other than bioconda development, and was brought up in #5134, bioconda/bioconda-utils#145 (comment), and #5355.

## Changes

- Simplified the steps to get started on recipes. All you need to start developing bioconda recipes is Python (2 or 3) and the PyYAML package. Optionally, docker. Everything else is handled automatically. This also simplifies setup on travis-ci.
- `~/.config/bioconda/conf.yml` stores the location of the bootstrapped installation so you don't have to specify each time
- `mulled-build` will now run by default, but there is an argument to disable it
- `--git-range` argument to `simulate-travis.py` specifies a git range to be used for both linting and building
- `--skip-linting` argument for . . . skipping linting
- Documentation in bioconda-utils has been updated to reflect these changes (specifically https://bioconda.github.io/contribute-a-recipe.html)

**Comment by sebastian-luna-valero** [comment]
Many thanks @daler !

**Comment by HenningTimm** [comment]
Thanks for the work @daler this makes my life a whole lot easier. This solves #5134 (and I will close it).

The installation and creation of the environment seem to work nicely. However I can not test if everything works due to another problem with docker (most likely related to #5027). I will create an Issue for this.

**Comment by jerowe** [comment]
@daler you just do the coolest things. ;-)

**Comment by peterjc** [comment]
Is this source warning expected under macOS?

```
$ python3 simulate-travis.py --bootstrap /tmp/miniconda 
Please run using "bash" or "sh", but not "." or "source"
/var/folders/x7/ygl010ps57g2s5q_l8y_jyjdmzsyc0/T/tmpn77rg0bp: line 18: return: can only `return' from a function or sourced script
PREFIX=/tmp/miniconda
installing: python-3.5.2-0 ...
installing: conda-env-2.6.0-0 ...
...
```

It seems to work, in that I can do:

```
$ python3 simulate-travis.py
Using config file /Users/xxx/.config/bioconda/conf.yml, which has the following contents:
{'CONDARC': '/tmp/miniconda/condarc', 'CONDA_ROOT': '/tmp/miniconda'}
+ bioconda-utils lint recipes config.yml --git-range master HEAD
17:18:42 BIOCONDA INFO Recipes newly unblacklisted:

17:18:42 BIOCONDA INFO Recipes to consider according to git: ...
...
```

**Comment by daler** [comment]
Thanks @peterjc I actually get it under linux as well. Fixed (at least for linux) in #5458, does this help for macOS?

**Comment by peterjc** [comment]
@daler Since it has been merged, testing #5458 was easy - I can confirm this also removes the warning on macOS:

```
$ rm -rf /tmp/miniconda/
$ python3 simulate-travis.py --bootstrap /tmp/miniconda 
PREFIX=/tmp/miniconda
installing: python-3.5.2-0 ...
...
```

**Comment by daler** [comment]
Great, thanks.
